Guided Tours: Add a theme sheet tour

mcsf opened the pull request:
This PR adds a theme sheet tour to show people around `/theme/:slug`.

Partly implements #7083 

![gt-themes-tour](https://cloud.githubusercontent.com/assets/150562/20775482/0e57809a-b753-11e6-8c68-3697dc0fb9f6.gif)

# To test

- Visit http://calypso.localhost:3000/theme/twentysixteen/?tour=themeSheetWelcomeTour and http://calypso.localhost:3000/theme/twentysixteen/MYSITE?tour=themeSheetWelcomeTour
- Go through the tour, make sure the experience makes sense and is fluid and that steps are rendered properly.

mcsf commented:
Things I think should change at the moment:
- More emphasis on the docs. Maybe point to the Features tag cloud at the bottom of the Overview tab?
- Likewise, if the theme has a Setup tab, let the user know that may be a way to unlock the full potential of the theme. Or something.
- The last step right now is prompting the user to go back to `/design`, and I think the experiences should be decoupled, so get rid of that one.

Suggestions? @marekhrabe @lsinger

lsinger commented:
> More emphasis on the docs.

Maybe just have a step that points to the "Overview" and "Support" tabs and says something to the tune of "Many themes come with powerful features. See here what this one can do."? 

> if the theme has a Setup tab

I think let's keep it simple for now and consider that in a future iteration / enhancement. 

> The last step right now is prompting the user to go back to `/design`, and I think the experiences should be decoupled, so get rid of that one.

Agree. The finish step could point to the "All Themes" link but not require users to click it, instead show a "We're all done" button. It *is* useful to know where to go back.

folletto commented:
> Docs / Setup

I'd add a single pointer that "hints" to both. Borrowing from @lsinger's copy is good.
Here's an idea:


* ( – baloon )
**This page shows all the details about a specific theme design. May I show you around?**
`[ Let's go! ]` `[ No thanks. ]`
* ( ↑ arrow )
**Here you can see the design in action in a demo site.**
`Click to continue.`
* ( ← arrow )
**This is the live demo. Take a look around, see if the design suits you! Then tap here to close.**
`Click to continue.`
* ( ↓ arrow )
**Many themes come with powerful features. See here what this one can do.**
`[ Next ]` `[ Close ]`
* ( ↑ arrow )
**This will activate the design you’re currently seeing on your site.**
`[ Next ]` `[ Close ]`
* ( ← arrow )
**If you want to see more themes, go back to the list here.**
`[ We're all done! ]`

How does it sound?

mcsf commented:
Sounds good!

> ( ← arrow )
> This is the live demo.

The positioning here hides the remaining icons of the preview toolbar (open ext. & change viewport). Should we try something else, like ↑ arrow? Also, should we hint at the viewport resizing feature?

folletto commented:
> The positioning here hides the remaining icons of the preview toolbar (open ext. & change viewport). Should we try something else, like ↑ arrow? 

Either we cover the icons, or we cover the top part of the site where there's usually title and logo. I don't feel either is a win, go for the one you feel is better.

> Also, should we hint at the viewport resizing feature?

The text is a bit long already. I'd say "yes", as long as we can do a shorter text. :)

mcsf commented:
Tour as of now:

![gt-themes-tour](https://cloud.githubusercontent.com/assets/150562/20775482/0e57809a-b753-11e6-8c68-3697dc0fb9f6.gif)

mcsf commented:
Rebased.

mcsf commented:
At this point I'm thinking that having an additional step for theme docs would be a burden. New users shouldn't have to look at docs at all; they'll read them whenever they're ready to have a more advanced grasp of a theme. Getting users to try them is the most important. ¿How's that sound?

lsinger commented:
> New users shouldn't have to look at docs at all; they'll read them whenever they're ready to have a more advanced grasp of a theme.

I still think it's important to make sure they *know* about them so they actually do look once they need them. I for one never learned about them and so had it much harder when I tried to get going on WordPress.com.

lsinger commented:
Just checked the tour on a mobile browser width. Things don't really work there. Maybe we need to add some scrolling to make sure all the things are in view? Maybe cf. #9602.

lsinger commented:
> Just checked the tour on a mobile browser width. Things don't really work there. Maybe we need to add some scrolling to make sure all the things are in view? Maybe cf. #9602.

Forget what I said. It's restricted to `isDesktop`. 😄 

Then again I can trigger it on a mobile width using the query arg. This seems like a fundamental thing we'll need to discuss. If HEs send around links to tour and users are on mobile ... what do we do? Maybe browser size is an important enough thing to warrant a specialized warning?

lsinger commented on client/layout/guided-tours/theme-sheet-tour.js:
```diff
@@ -0,0 +1,133 @@
+			isEnabled( 'guided-tours/theme' ),
+			isNewUser,
+			isDesktop
+			) }
```
add 50/50 A/B test as last condition?

folletto commented:
> New users shouldn't have to look at docs at all; they'll read them whenever they're ready to have a more advanced grasp of a theme. Getting users to try them is the most important. ¿How's that sound?

We recently introduced links from both Editor and Customizer to the support page because people missed it... as well as tests highlight it's something that is easily overlooked so... I'd try to have that step instead of the live demo one ^^'

To be clear: I agree they _shouldn't_, but we aren't there yet.

mcsf commented:
> Then again I can trigger it on a mobile width using the query arg. This seems like a fundamental thing we'll need to discuss. If HEs send around links to tour and users are on mobile ... what do we do? Maybe browser size is an important enough thing to warrant a specialized warning?

Agree, a pertinent question. Not sure what the fix is. In the meantime, I changed the `isDesktop` requirement into a `not( isMobile )`, thus allowing more users in. The whole thing is too shaky on mobile, and I'd rather make sure it's a robust experience before enabling it on mobile, and that sort of work isn't part of the immediate future.

> We recently introduced links from both Editor and Customizer to the support page because people missed it... as well as tests highlight it's something that is easily overlooked so... I'd try to have that step instead of the live demo one ^^'

Added a docs step.

lsinger commented:
Rebased.

lsinger commented:
The "click live demo" and "activate here" steps would need to scroll if width < 940, but they don't. That means that the tour seems quit / suspended until one scrolls to the properly place manually. 

Screen capture:

![themes-no-scroll](https://cloud.githubusercontent.com/assets/23619/20926926/5aa270a2-bbbf-11e6-8b07-d11bfd6387c7.gif)

lsinger commented:
The "click here for live demo" and "this is where you would activate" steps aren't visible and aren't scrolled to if width < 960px. Tour seems quit / suspended until one scrolls manually. Screen capture:

![themes-no-scroll](https://cloud.githubusercontent.com/assets/23619/20926991/b9c30c36-bbbf-11e6-88df-8993b225de8f.gif)

lsinger commented on client/layout/guided-tours/theme-sheet-welcome-tour.js:
```diff
@@ -0,0 +1,153 @@
+} from 'state/ui/guided-tours/contexts';
+import { isDesktop, isMobile } from 'lib/viewport';
+
+const PickActivateStep = () => (
```
Rename? It isn't a full step, but step text / description. `PickActivateStepText`?

lsinger commented on client/layout/guided-tours/theme-sheet-welcome-tour.js:
```diff
@@ -0,0 +1,153 @@
+const PickActivateStep = () => (
+	<p>
+		{ translate(
+			'This will activate the design you’re currently seeing on your site.'
```
I wasn't aware we're using `’`. Shouldn't we use `'` everywhere for consistency? And then use double quotes here for quoting the string?

lsinger replied:
"This would activate the theme for your site."

mcsf replied:
Ah, that was accidental — good catch! Probably a copy-paste from somewhere with auto-quotes.

lsinger commented on client/layout/guided-tours/theme-sheet-welcome-tour.js:
```diff
@@ -0,0 +1,153 @@
+);
+
+export const ThemeSheetWelcomeTour = makeTour(
+	<Tour name="theme"
```
Rename to `themeSheetWelcomeTour` here as well?

mcsf replied:
Ah, right! 😅

lsinger commented on client/layout/guided-tours/theme-sheet-welcome-tour.js:
```diff
@@ -0,0 +1,153 @@
+		<Step name="init" placement="right" next="live-preview">
+			<p>
+				{ translate(
+					'This page shows all the details about a specific theme ' +
```
I don't think we say "theme design" anywhere? Shouldn't we just say "theme"?

mcsf commented on config/development.json:
```diff
@@ -43,6 +43,7 @@
 		"dev/test-helper": true,
 		"guided-tours": true,
 		"guided-tours/main": true,
+		"guided-tours/theme": true,
```
I'm going to change this too to `theme-sheet-welcome` and enable it in `wpcalypso` as well.

mcsf commented:
Addressed the feedback, rebased against `master` after the #9602 merge.

mcsf commented:
> The "click live demo" and "activate here" steps would need to scroll if width < 940, but they don't. That means that the tour seems quit / suspended until one scrolls to the properly place manually.

Yeah. 😞 Not sure what this fix should be there. Should we simply require `isDesktop`?

lsinger commented:
> Should we simply require `isDesktop`?

Yes, let's go with that for now.

lsinger commented:
LGTM! :shipit:
